presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java

```diff
@@ -194,41 +206,42 @@ public synchronized boolean appendToNext(PageBuilder pageBuilder, int outputChan
         }
     }
 
+    /**
+     * Synchronization in commit() is required because each writing thread must have
+     * memory barrier, so that it writes its state of visitedPositions to main memory.
+     *
+     * Synchronization in getVisitedPositions() is to make sure that reader sees the
+     * latest version of the visitedPositions
+     */
     @ThreadSafe
     private static class OuterPositionTracker
     {
-        private final List<Supplier<LookupSource>> lookupSourceSuppliers;
-
-        @GuardedBy("this")
         private final boolean[][] visitedPositions;
+        private boolean committed;
 
-        @GuardedBy("this")
-        private boolean finished;
-
-        public OuterPositionTracker(List<Supplier<LookupSource>> lookupSourceSuppliers)
+        public OuterPositionTracker(List<Integer> positionCounts)
         {
-            this.lookupSourceSuppliers = lookupSourceSuppliers;
-            visitedPositions = new boolean[lookupSourceSuppliers.size()][];
-            for (int source = 0; source < lookupSourceSuppliers.size(); source++) {
-                try (LookupSource lookupSource = lookupSourceSuppliers.get(source).get()) {
-                    visitedPositions[source] = new boolean[lookupSource.getJoinPositionCount()];
-                }
+            visitedPositions = new boolean[positionCounts.size()][];
+            for (int partition = 0; partition < visitedPositions.length; partition++) {
+                visitedPositions[partition] = new boolean[positionCounts.get(partition)];
             }
         }
 
-        public synchronized void positionVisited(int partition, int position)
+        // no synchronization here, because it would be very expensive
+        public void positionVisited(int partition, int position)
         {
-            verify(!finished);
             visitedPositions[partition][position] = true;
         }
 
-        public synchronized OuterPositionIterator getOuterPositionIterator()
+        public synchronized void commit()
+        {
+            committed = true;
+        }
+
+        public synchronized boolean[][] getVisitedPositions()
```

Comment:
I do not see how this code enforces that
getVisitedPositions is called only after _every_ other thread went through synchronized section on `OuterPositionTracker`.
Lets discuss this offline.